builtin/logical/transit/path_decrypt.go

```diff
@@ -84,24 +109,115 @@ func (b *backend) pathDecryptWrite(
-		switch err.(type) {
-		case errutil.UserError:
-			return logical.ErrorResponse(err.Error()), logical.ErrInvalidRequest
-		case errutil.InternalError:
-			return nil, err
-		default:
-			return nil, err
+	batchInputRaw := d.Get("batch").(string)
+	var batchInputArray []BatchDecryptionItemRequest
+	var batchInput []byte
+	if len(batchInputRaw) != 0 {
+		batchInput, err = base64.StdEncoding.DecodeString(batchInputRaw)
+		if err != nil {
+			return logical.ErrorResponse("failed to base64-decode batch input"), logical.ErrInvalidRequest
+		}
+
+		if err := jsonutil.DecodeJSON([]byte(batchInput), &batchInputArray); err != nil {
+			return nil, fmt.Errorf("invalid input: %v", err)
+		}
+
+	} else {
+		ciphertext := d.Get("ciphertext").(string)
+		if len(ciphertext) == 0 {
+			return logical.ErrorResponse("missing ciphertext to decrypt"), logical.ErrInvalidRequest
+		}
+
+		batchInputArray = make([]BatchDecryptionItemRequest, 1)
+		batchInputArray[0] = BatchDecryptionItemRequest{
+			Ciphertext: ciphertext,
+			Context:    d.Get("context").(string),
+			Nonce:      d.Get("nonce").(string),
 		}
 	}
 
-	// Generate the response
-	resp := &logical.Response{
-		Data: map[string]interface{}{
-			"plaintext": plaintext,
-		},
+	if len(batchInputArray) == 0 {
+		return logical.ErrorResponse("missing input to process"), logical.ErrInvalidRequest
+	}
+
+	contextSet := true
+
+	for _, item := range batchInputArray {
+		if item.Context == "" && contextSet {
+			contextSet = false
+		}
+
+		if item.Context != "" && !contextSet {
+			return logical.ErrorResponse("context should be set either in all the request blocks or in none"), logical.ErrInvalidRequest
+		}
+	}
+
+	batchResponseItems := make([]BatchDecryptionItemResponse, len(batchInputArray))
+	for i, item := range batchInputArray {
+		if item.Ciphertext == "" {
+			batchResponseItems[i] = BatchDecryptionItemResponse{
+				Error: "missing ciphertext to decrypt",
+			}
+			continue
+		}
+
+		var itemContext []byte
+		if len(item.Context) != 0 {
+			itemContext, err = base64.StdEncoding.DecodeString(item.Context)
+			if err != nil {
+				batchResponseItems[i] = BatchDecryptionItemResponse{
+					Error: "failed to base64-decode context",
+				}
+				continue
+			}
+		}
+
+		var itemNonce []byte
+		if len(item.Nonce) != 0 {
+			itemNonce, err = base64.StdEncoding.DecodeString(item.Nonce)
+			if err != nil {
+				batchResponseItems[i] = BatchDecryptionItemResponse{
+					Error: "failed to base64-decode nonce",
+				}
+			}
+			continue
+		}
+
+		plaintext, err := p.Decrypt(itemContext, itemNonce, item.Ciphertext)
+		if err != nil {
+			switch err.(type) {
+			case errutil.UserError:
+				return logical.ErrorResponse(err.Error()), logical.ErrInvalidRequest
+			case errutil.InternalError:
+				return nil, err
+			default:
+				return nil, err
+			}
+		}
+
+		batchResponseItems[i] = BatchDecryptionItemResponse{
+			Plaintext: plaintext,
+		}
+	}
+
+	resp := &logical.Response{}
+	if len(batchInputRaw) != 0 {
```

Comment:
I think here it'd be a bit nicer to key off `if len(batchResponseItems) > 1` since otherwise you have to context switch the initial raw input into your mind.